src/golang/cmd/server/handler/connect_integration.go

```diff
@@ -202,58 +207,241 @@ func (h *ConnectIntegrationHandler) Perform(ctx context.Context, interfaceArgs i
 		// This integration should be used as the new storage layer.
 		// In order to do so, we need to migrate all content from the old store
 		// to the new store. This requires pausing the server and then restarting it.
-		// All of this logic is performed asynchronously so that the user knows that
-		// the connect integration request has succeeded and that the migration is now
-		// under way.
+
+		// The migration logic is performed asynchronously, and it's process is tracked
+		// as a new entry in the `storage_migration` table:
+		storageMigrationObj, err := h.StorageMigrationRepo.Create(
+			ctx,
+			&integrationObj.ID,
+			h.Database,
+		)
+		if err != nil {
+			return emptyResp, http.StatusInternalServerError, errors.Wrap(err, "Unable to migrate storage.")
+		}
+
+		// If the migration is successful, the new entry is given a success execution status, along with `current=True`.
+		// If the migration is unsuccessful, the error is recorded on the new entry in `storage_migration`.
 		go func() {
+			// Shadows the context in the outer scope on purpose.
+			ctx := context.Background()
+
 			log.Info("Starting storage migration process...")
 			// Wait until the server is paused
 			h.PauseServer()
 			// Makes sure that the server is restarted
 			defer h.RestartServer()
 
-			// Wait until there are no more workflow runs in progress
-			lock := utils.NewExecutionLock()
-			if err := lock.Lock(); err != nil {
-				log.Errorf("Unexpected error when acquiring workflow execution lock: %v. Aborting storage migration!", err)
-				return
-			}
+			execState := storageMigrationObj.ExecState
+
+			var err error
 			defer func() {
-				if err := lock.Unlock(); err != nil {
-					log.Errorf("Unexpected error when unlocking workflow execution lock: %v", err)
+				if err != nil {
+					// Regardless of whether the migration succeeded, we should track the finished timestamp.
+					finishedAt := time.Now()
+					execState.Timestamps.FinishedAt = &finishedAt
+
+					execState.UpdateWithFailure(
```

Comment:
Interesting, so we are not updating timestamp in this `UpdateWithFailure` call?

Reply:
We are not. Do you think we should?

Reply:
Actually just went ahead and did it.